sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/EquivalentExpressions.scala

```diff
@@ -130,7 +133,19 @@ class EquivalentExpressions {
   //   2. ConditionalExpression: use its children that will always be evaluated.
   private def childrenToRecurse(expr: Expression): Seq[Expression] = expr match {
     case _: CodegenFallback => Nil
-    case c: ConditionalExpression => c.alwaysEvaluatedInputs
+    case c: ConditionalExpression =>
+      if (shortcut) {
+        // The subexpression in conditional expression may not need to eval even if it appears
```

Comment:
why only in conditional expression? `where or(a, and(b, b))` has the same problem, right?

Reply:
IIUC, CSE only supports project and aggregate. `Predicate` seems not likely appear without conditional expr.

Reply:
not likely, but possible, `select or(a, and(b, b))` is also valid.

BTW, I think we should not enable this new change by default, as it may lead to perf regression.

Reply:
make sense, addressed ！